Exposing fontprovider

Jjastiny opened the pull request:
### Checklist
- [ ] If applicable, create follow-up issues for `purchases-ios` and hybrids

### Motivation
- It's not possible for us to change the font family in react-native-purchase.
- My hope was to expose this font provider methods for the RN Bridge so that we can at least set custom font family.

### Description
- Exposing fontProviders for PaywallFooterView and PaywallView.
- ~~Invalidate views when font provider changes.~~ theme composition local provider should update this autmoatically.

Need to add something like this on the React Native side.
If user did link the font via
```xml
<font-family xmlns:app="http://schemas.android.com/apk/res/android">
        <font app:fontStyle="normal" android:fontWeight="200" android:font="@font/montserrat" />
</font-family>
```

```kotlin
    ReactFontManager.getInstance().addCustomFont(this, "Montserrat", R.font.montserrat);
```

We could just bring the font family 
```kotlin
    @ReactProp(name= "fontFamily")
    fun setFontFamily(view: PaywallFooterView, fontFamily: String?){
        val typeface = ReactFontManager.getInstance()
            .getTypeface(fontFamily /** Montserrat in this case. **/, ReactFontManager.TypefaceStyle.NORMAL, assets)
        
        view.setFontProvider(CustomFontProvider(FontFamily(typeface)))
    }
```

<img width="350" alt="Screenshot 2024-01-26 at 10 04 03 PM" src="https://github.com/RevenueCat/purchases-android/assets/6516487/e8e41b0d-c1fd-4446-8d5b-11cb9de200e5">

Jjastiny committed:
Expose font provider, so that react native bridge can redraw with custom font.

Jjastiny committed:
Add comments. Expose fontProvider for paywallView as well.

Jjastiny committed:
remove unnecessary invalidate from setFontProvider.

Jjastiny committed:
making the FontProvider and OfferingId to a mutableState so that composeview can listen to the state change.

tonidero commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/views/PaywallFooterView.kt:
```diff
@@ -85,35 +89,43 @@ open class PaywallFooterView : FrameLayout {
+                        PaywallOptions.Builder { dismissHandler?.invoke() }
+                            .setListener(internalListener)
+                            .setFontProvider(fontProviderState)
+                            .setOfferingId(offeringIdState)
```
I may be missing something, but I believe this won't recompute the options on a change on the `offeringIdState`. 
The inside of the `remember` function is not a mutableState, so I don't think it would be called once some mutableState dependants change (Do lmk if I'm wrong!)

Jjastiny replied:
@tonidero 
so any changes in the state within the key in `remember(key)` should trigger this lambda again with the new PaywallOptions.

tonidero replied:
@Jjastiny From what I understand, the next time it recomposes, it would indeed create a new PaywallOptions instance with the new value due to that `remember(key)`. However, it wouldn't trigger a recomposition by itself since nothing in the compose context is listening to the changes of a `mutableState` (I believe being part of the key of the `remember` does not trigger a recomposition).

Having said that, I haven't tried this code, but I believe the change wouldn't be triggered. Have you been able to confirm the change triggers with this?

Jjastiny replied:
@tonidero 
yeap ive confirmed that this does indeed cause recomposition; however,the code that was merged to main in ur main branch does not. (It only recomposes once) 😱I think it may cause future bugs, so it would be great if we could start looking at it again!

Jjastiny replied:
yeah i stand corrected, it looks like it's recomposing in the main branch!

Jjastiny committed:
Merge branch 'main' of other.github.com:Jjastiny/purchases-android into exposing-fontprovider

# Conflicts:
#	ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/views/PaywallFooterView.kt

Jjastiny committed:
use abstract compose view, to remove unnecessary addSubView(ComposeView(context))
expose the font provider. set a private getter and setter to reduce some boilerplate

Jjastiny committed:
remove unused imports

Jjastiny committed:
remove paywall view font provider for now

Jjastiny commented:
@tonidero How do you feel about adding this PR to the purchase-android?
Is there a roadmap for font provider setter? If so, I can wait 😄

tonidero commented:
Hey @Jjastiny, I've just tested these changes and they work great! 

We don't have a timeline yet for supporting custom fonts on the hybrids, but it shouldn't take too long we hope, we are actively working on better support there.

I think this PR can be merged once we apply the same changes to the `PaywallView` (sorry for the duplication 😢). Do you mind applying the changes there and open a PR against our repo? Then we can get it merged.

Jjastiny commented:
@tonidero 
I do not mind at all!
If you dont mind, I will create ios and react-native bridge changes too :)

tonidero commented:
Sure thing! Contributions are more than welcome 😄 thank you!!

Jjastiny committed:
PaywallView now uses AbstractComposeView, to abstract the composable function out. Expose fontProvider.

Jjastiny commented:
Closing. as it was merged to purchases-android.
